Testing mpl images for graphs

singhmeet11 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
As of now there is no testing module for graph images. This issue will be resolved by this PR and is motivated from issue #6397.  


### Details and comments
List of changes and additions for creating the tests for graphs.
1. There was no option for saving graph images like we can save images of circuit using the parameter `filename`.  Made changes in  `qiskit/visualization/state_visualization.py`  so that we can save `.png` using `state_drawer` . (This has only been implemented for `plot_bloch_multivector` for now).
2. Made different folders for `Graphs` and `Circuit` in `qiskit-terra\test\ipynb\mpl` . These house the reference images along with the `testing.py` files. Made this for clarification and because this resolved an issue I was facing with  `results.py` and `mpl_results_graph.py`.
3. `mpl_results_graph.py` is identical to `results.py` with a few modifications for the file names.
4. `test_graph_matplotlib_drawer.py` works similar to `test_circuit_matplotlib_drawer.py`. Originally I changed it a bit, but with the modifications to `state_visualization.py` this works properly.

The test works properly when run via Binder - 

![new_test](https://user-images.githubusercontent.com/76438148/124100961-8a898400-da7c-11eb-8804-dc56768eab70.PNG)

The added test for `plot_bloch_multivector` can be seen above

javabster commented on qiskit/visualization/state_visualization.py:
```diff
@@ -233,7 +233,7 @@ def plot_bloch_vector(bloch, title="", ax=None, figsize=None, coord_type="cartes
 
 @deprecate_arguments({"rho": "state"})
-def plot_bloch_multivector(state, title="", figsize=None, *, rho=None, reverse_bits=False):
+def plot_bloch_multivector(state, filename=None, title="", figsize=None, *, rho=None, reverse_bits=False):
```
New params should generally be added at the end of the params list :)

singhmeet11 replied:
done

javabster commented on test/ipynb/mpl_results_graph.py:
```diff
@@ -0,0 +1,200 @@
+# This code is part of Qiskit.
```
I'm confused about why a whole new file was created for this? There is a lot of duplicated code here, I think it would be better to try integrate the graph results into the existing `mpl_results.py` file if possible

singhmeet11 replied:
@javabster I tried to actually use only one file `results.py` but because I am creating two files inside the `mpl` folder( `circuit` and `graph`) and the path to specific folders inside these have to be provided in order for the file to work properly, I was not able to integrate the whole code into one file. 
How about I remove the folders `graph` and `circuit` and store all the data in the same `.json` file and same `references` folder(like there won't be any different references and .py files for graph and circuit images). Then only one file will be able to do all the work.

javabster replied:
I think it can be worked into one file using the specific file names required, I will push a commit shortly to show you what I mean 😄

javabster replied:
Ok I've pushed some changes and it looks like it works now with just the original `results.py` so this file can be deleted

singhmeet11 replied:
Yeah, thanks I was kinda stuck there.

javabster commented on test/ipynb/mpl_tester.ipynb:
```diff
@@ -2,24 +2,133 @@
  "cells": [
```
There seems to be a lot of extra metadata added to this file which we probably don't want to merge in. Could you please commit a version of this file that hasn't been run and just includes the changes we want?

enavarro51 replied:
@singhmeet11 This is a common problem when using the mpl_tester.ipynb on a local system. If you test with it, data from your local gets written back into it. What I do is copy the mpl_tester.ipynb from main to another name and place it outside of qiskit-terra on my local. Then I copy it into qiskit-terra/test/ipynb to do the testing, but before doing a commit, I delete the copy from there.

singhmeet11 replied:
Thanks for your suggestion I followed it and it mpl_tester.ipynb looks okay now

javabster commented on test/ipynb/results.py:
```diff
@@ -146,16 +147,16 @@ def summary(self):
 
         if len(self.mismatch) >= 2:
-            Results._zipfiles(self.mismatch, "mpl/mismatch.zip")
+            Results._zipfiles(self.mismatch, "mpl/Circuit/mismatch.zip")
```
Folder names should be lowercase `circuit` and `graph` please 😊

singhmeet11 replied:
done

javabster reviewed (commented):
Thanks for the work so far @singhmeet11! I've left a few comments 😄

singhmeet11 commented:
@javabster sorry I forgot to add the link to this PR and I will look into the comments that you have posted in a day or two. Thanks

javabster commented on qiskit/visualization/state_visualization.py:
```diff
@@ -1239,6 +1244,7 @@ def state_drawer(state, output=None, **drawer_args):
             default_output = config.get("state_drawer", "repr")
         output = default_output
     output = output.lower()
+    # filename = drawer_args.pop('filename')
```
can you remove comments that are no longer needed please 😄

javabster reviewed (commented):
This is looking good, just a few things to tidy up then should be good to go

singhmeet11 commented:
@javabster I think the basic architecture works fine, I will now start adding the test for other cases as well, these test will of course be very similar to ` test_plot_bloch_multivector.` Do you have any suggestions about this ??

javabster commented:
> @javabster I was actually thinking that we could expand this for other cases as well(other plots). Would that be okay ?? I don't think it would take a lot of time considering that the basic architecture works fine.

sure go for it!

singhmeet11 commented:
@javabster So I have successfully added the tests for `paulivec, state_city, Bloch sphere`. But I am facing issues with the `qsphere`, the tests run normally on my laptop but when I run the same tests on binder they fail. The error

```
ERROR: test_plot_state_qsphere (__main__.TestGraphMatplotlibDrawer)
test for plot_state_qsphere
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/conda/envs/notebook/lib/python3.7/site-packages/qiskit/visualization/state_visualization.py", line 765, in plot_state_qsphere
    import seaborn as sns
ModuleNotFoundError: No module named 'seaborn'
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/jovyan/test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py", line 132, in test_plot_state_qsphere
    self.graph_drawer(state=state, output="qsphere", filename="qsphere.png")
  File "/home/jovyan/test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py", line 77, in wrapper
    results = func(*args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.7/site-packages/qiskit/visualization/state_visualization.py", line 1304, in state_drawer
    return draw_func(state, **drawer_args)
  File "/srv/conda/envs/notebook/lib/python3.7/site-packages/qiskit/utils/deprecation.py", line 62, in wrapper
    return func(*args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.7/site-packages/qiskit/visualization/state_visualization.py", line 771, in plot_state_qsphere
    ) from ex
qiskit.exceptions.MissingOptionalLibraryError: "The 'seaborn' library is required to use 'plot_state_qsphere'. You can install it with 'pip install seaborn'."
```

So basically the environment that binder is running needs seaborn (I actually used pip to install it and the .ipynb file runs fine) But I don't think adding pip install in the mpl_circuit file will look clean so where do I add the pip install command for seaborn?

The other problem- I also added a test for `histogram` but after each run, the probabilities change after each run, due to noise. Now, I am looking into how to remove noise completely. I actually tried using the Noise model and equate the probability of error = 0 but that did not work.
Let me know your thoughts on these and thanks for your help.

singhmeet11 commented:
@javabster I have added tests for the 6 main plots, `qsphere`, `paulivec`, `state_city`, `histogram`, `bloch_vector` and finally `hinton` . I don't think there are any other major ones.
One thing the `histogram` plot is made using `counts = {"11": 500, "00": 500}` basically I fixed the outcome because executing it leads to fluctuation of counts. Originally thought that this was happening due to noise but https://quantumcomputing.stackexchange.com/questions/9813/how-to-run-the-qiskit-aer-simulator-without-noise shows that the reason for this is different and it is not removable. Moreover I don't think the purpose of this test is to test the simulator and because this will tell if a change has happened in the `plot_hisotgram` command , I think this is good enough. 
Let me know what you think

![qiskit_snip](https://user-images.githubusercontent.com/76438148/125121750-fe2d3000-e111-11eb-9566-8ffcf00008a2.PNG)
Here is a snapshot of the snapshot tests

enavarro51 commented:
@singhmeet11 This looks nice. I like the white space separator between the circuit images and the visualizations. And I'd agree you're not trying to catch simulator problems, just the visuals.

singhmeet11 commented:
@javabster The tests run fine too now. Yayy

javabster commented on postBuild:
```diff
@@ -9,6 +9,7 @@
 # - appmode: jupyter extension for executing the notebook
```
```suggestion
# - appmode: jupyter extension for executing the notebook
# - seaborn: visualisation pacakge required for some graphs
```

javabster reviewed (approved):
This is great work thanks @singhmeet11! Just one tiny addition, otherwise lgtm 🚀

singhmeet11 commented:
@javabster Thanks! I think this is ready to be merged now.

javabster commented:
> @javabster Thanks! I think this is ready to be merged now. Wait, what does dismissing stale review mean ??

it just means that the commit you pushed mean my previous comment was outdated. Nothing to worry about 😄

1ucian0 commented on postBuild:
```diff
@@ -7,8 +7,10 @@
 # - pylatexenc: for MPL drawer
 # - pillow: for image comparison
 # - appmode: jupyter extension for executing the notebook
+# - seaborn: visualisation pacakge required for some graphs
 pip install matplotlib pylatexenc pillow appmode
 pip install .
+pip install seaborn
```
```suggestion
# - seaborn: visualisation pacakge required for some graphs
pip install matplotlib pylatexenc pillow appmode seaborn
pip install .
```

1ucian0 reviewed (approved):
This is great! Thanks!
